internal/filter/parser.y

```diff
@@ -0,0 +1,183 @@
+// The second argument `rule` is supposed to be a filter.Any *Chain contains the first two conditions.
+// We then call this function when the parser is processing the logical `&` op and the Unlike condition,
+// and what this function will do is logically re-group the conditions into `foo=bar|(bar~foo&col!~val)`.
+func reduceFilter(op string, rule Filter, rules ...Filter) Filter {
+	chain, ok := rule.(*Chain);
+	if ok && chain.op == Any && LogicalOp(op) == All {
+	    // Retrieve the last pushed condition and append it to the new "And" chain instead
+	    andChain, _ := NewChain(All, chain.pop())
+	    andChain.add(rules...)
+
+	    chain.add(andChain)
+
+	    return chain
+	}
+
+	// If the given operator is the same as the already existsing chains operator (*chain),
+	// we don't need to create another chain of the same operator type. Avoids something
+	// like &Chain{op: All, &Chain{op: All, ...}}
+	if chain == nil || chain.op != LogicalOp(op) {
+	    newChain, err := NewChain(LogicalOp(op), rule)
+	    if err != nil {
+		// Just panic, filter.Parse will try to recover from this.
+		panic(err)
+	    }
+
+	    chain = newChain
+	}
+
+	chain.add(rules...)
+
+	return chain
+}
+%}
+
+%union {
+	expr Filter
+	text string
+}
+
+%type <expr> filter_rule
+%type <expr> filter_chain
+%type <expr> conditions_expr
+%type <expr> maybe_negated_condition_expr
+%type <expr> condition_expr
+%type <expr> exists_expr
+
+%type <text> comparison_op
+%type <text> optional_negation
+%type <text> identifier
+%type <text> logical_op
+
+%token T_EQUAL "="
+%token T_UNEQUAL "!" T_EQUAL
+%token T_LIKE "~"
+%token T_UNLIKE "!" T_LIKE
+%token T_LESS_THAN "<"
+%token T_GREATER_THAN ">"
+%token T_LESS_THAN_OR_EQUAL "<" T_EQUAL
+%token T_GREATER_THAN_OR_EQUAL ">" T_EQUAL
+%token <text> T_STRING
+%token <text> T_IDENTIFIER
+
+%type <text> T_EQUAL
+%type <text> T_UNEQUAL
+%type <text> T_LIKE
+%type <text> T_UNLIKE
+%type <text> T_LESS_THAN
+%type <text> T_GREATER_THAN
+%type <text> T_LESS_THAN_OR_EQUAL
+%type <text> T_GREATER_THAN_OR_EQUAL
+
+%type <text> "!" "&" "|"
+
+// This is just used for declaring explicit precedence and resolves shift/reduce conflicts
+// in `filter_chain` and `conditions_expr` rules.
+%nonassoc PREFER_SHIFTING_LOGICAL_OP
+
+%nonassoc T_EQUAL T_UNEQUAL T_LIKE T_UNLIKE
+%nonassoc T_LESS_THAN T_LESS_THAN_OR_EQUAL T_GREATER_THAN T_GREATER_THAN_OR_EQUAL
+
+%left "!" "&" "|"
+%left "("
+%right ")"
+
+%%
+
+filter_rule: filter_chain logical_op filter_chain
+	{
+		$$ = reduceFilter($2, $1, $3)
+		yylex.(*Lexer).rule = $$
+	}
+	| filter_chain
+	{
+		yylex.(*Lexer).rule = $$
+	}
+	;
+
+filter_chain: conditions_expr logical_op maybe_negated_condition_expr
+	{
+		$$ = reduceFilter($2, $1, $3)
+	}
+	| conditions_expr %prec PREFER_SHIFTING_LOGICAL_OP
+	;
+
+conditions_expr: maybe_negated_condition_expr logical_op maybe_negated_condition_expr
+	{
+		$$ = reduceFilter($2, $1, $3)
+	}
+	| maybe_negated_condition_expr %prec PREFER_SHIFTING_LOGICAL_OP
+	;
```

Comment:
Am I missing something or does this limit a `filter_chain` to at most 3 `condition_expr`?

Reply:
Yes, it does, which limits `filter_rule` to 6 `condition_expr`, so parsing `a & b & c & d & e & f & g` fails (spaces for readability, parsing fails with and without them).

Reply:
Yes, I noticed that last time too, but didn't have time to push all the fixes.

Reply:
Should be fixed with https://github.com/Icinga/icinga-notifications/pull/41/commits/5484f912d2498beaee1182f23d6dbcdceab82573